### ISSUE TITLE
Investigate latency issue with connector builder

### ISSUE THREAD

**Issue by girarda** [airbytehq/airbyte#25632]
- [ ] What is the root cause?
- [ ] Can we add instrumentation to understand where we spend time?

**Comment by girarda** [comment]
We decided to timebox this issue, but I'll leave it open because the root cause isn't identified.

This flamegraph shows that a lot of the time is spent in getMessagesByType, but the time includes
- Running the Python
- Reading from the Python process
- Parsing the process' output

We don't have great telemetry on the Python-side of the flow.

![Screen Shot 2023-04-27 at 2 38 44 PM (1) (1)](https://user-images.githubusercontent.com/1451943/235278310-0581e4d2-6cde-47da-a390-f8e7094e882d.png)

Potential next steps
- Deploy the traces to Cloud and hope we get more insights
- Add instrumentation to the Python process. One option would be to emit trace messages

**Comment by girarda** [comment]
From the flamegraph, it doesn't seem like the issue is with serialization because most of the time is spent on File I/O, which I think means the thread was waiting on the Python process.
<img width="848" alt="Screen Shot 2023-05-02 at 7 45 33 AM" src="https://user-images.githubusercontent.com/1451943/235701411-7e92d84d-0ec0-4cf9-80e3-9e05232270f5.png">

**Comment by girarda** [comment]
Some numbers of the execution time of the python process:
- It takes 0.16-0.18s to create the source because of the manifest validation

```
2023-05-02 18:31:44 INFO i.a.w.i.VersionedAirbyteStreamFactory(toAirbyteMessage):369 - Function '_get_args' executed in 0.0000s
2023-05-02 18:31:44 INFO i.a.w.i.VersionedAirbyteStreamFactory(toAirbyteMessage):369 - Function 'get_config_and_catalog_from_args' executed in 0.0014s
2023-05-02 18:31:44 INFO i.a.w.i.VersionedAirbyteStreamFactory(toAirbyteMessage):369 - Function 'get_limits' executed in 0.0000s
2023-05-02 18:31:44 INFO i.a.w.i.VersionedAirbyteStreamFactory(toAirbyteMessage):369 - Function '_validate_source' executed in 0.1831s
2023-05-02 18:31:44 INFO i.a.w.i.VersionedAirbyteStreamFactory(toAirbyteMessage):369 - Function 'create_source' executed in 0.1834s
2023-05-02 18:31:44 INFO i.a.w.i.VersionedAirbyteStreamFactory(toAirbyteMessage):369 - Function 'list_streams' executed in 0.0028s
2023-05-02 18:31:44 INFO i.a.w.i.VersionedAirbyteStreamFactory(toAirbyteMessage):369 - Function 'handle_connector_builder_request' executed in 0.0028s
2023-05-02 18:31:44 INFO i.a.w.i.VersionedAirbyteStreamFactory(toAirbyteMessage):369 - Function 'handle_request' executed in 0.1877s
2023-05-02 18:31:44 INFO i.a.w.i.VersionedAirbyteStreamFactory(toAirbyteMessage):369 - Function '_main' executed in 0.1877s
```

**Comment by girarda** [comment]
I ran the profiler on "just running the list stream command" in a loop and time it

```
      ApplicationBeanFactory factory = new ApplicationBeanFactory();
      SynchronousCdkCommandRunner commandRunner = factory.synchronousPythonCdkCommandRunner();
      AirbyteRecordMessage output = commandRunner.runCommand("list_streams", config, "");
```

<img width="978" alt="Screen Shot 2023-05-02 at 4 43 18 PM" src="https://user-images.githubusercontent.com/1451943/235808128-affcb250-3098-4494-9190-d5b941155cfb.png">

- Most of the time is spent waiting for the Python output

I added instrumentation in the Python process, and it only used ~0.13s out of 0.477s of the full flow, which seems to indicate a significant amount of time is spent actually starting the Python process.

This seems at odd with what I said in [this comment](https://github.com/airbytehq/airbyte/issues/25632#issuecomment-1531612290), but [this stackoverflow thread](https://stackoverflow.com/questions/41891137/java-large-startup-overhead-of-external-application-processbuilder-nuprocess) mentions that starting a process from Java takes more time than starting it from the command line.

**Comment by girarda** [comment]
still something that would be worth doing eventually, but we should re-open a new issue with a more targeted goal
